### PR TITLE
Fixed #26888 -- Fixed concurrency issue in URL resolver.

### PR DESCRIPTION
**Pull request by knbk**
Fixed a regression in 625b8e9.

Fixed improper short-circuiting that could lead to a KeyError when threads
concurrently call RegexURLResolver._populate().

### CODE REVIEW

File: django/urls/resolvers.py [hunk 1]
```diff
@@ -178,9 +179,13 @@ def __repr__(self):
         )
 
     def _populate(self):
-        if self._populating:
+        # Short-circuit if called recursively in this thread to prevent
+        # infinite recursion. Concurrent threads may call this at the same
+        # time, and will need to continue, so we set 'populating' on a
```
Comment (timgraham) [inline django/urls/resolvers.py]: chop first comma and "we"

File: tests/urlpatterns_reverse/tests.py [hunk 2]
```diff
@@ -429,6 +430,18 @@ def test_namespaced_view_detail(self):
         self.assertTrue(resolver._is_callback('urlpatterns_reverse.nested_urls.View3'))
         self.assertFalse(resolver._is_callback('urlpatterns_reverse.nested_urls.blub'))
 
+    def test_populate_concurrency(self):
+        """
+        Verifies that RegexURLResolver._populate() can be called
```
Comment (timgraham) [inline tests/urlpatterns_reverse/tests.py]: chop "Verifies that"

File: tests/urlpatterns_reverse/tests.py [hunk 3]
```diff
@@ -429,6 +430,18 @@ def test_namespaced_view_detail(self):
...
+        concurrently (#26888).
```
Comment (MarkusH) [inline tests/urlpatterns_reverse/tests.py]: "... concurrently but not more than once per thread ..."

File: tests/urlpatterns_reverse/tests.py [hunk 4]
```diff
@@ -429,6 +430,17 @@ def test_namespaced_view_detail(self):
         self.assertTrue(resolver._is_callback('urlpatterns_reverse.nested_urls.View3'))
         self.assertFalse(resolver._is_callback('urlpatterns_reverse.nested_urls.blub'))
 
+    def test_populate_concurrency(self):
+        """
+        RegexURLResolver._populate() can be called concurrently (#26888).
```
Comment (MarkusH) [inline tests/urlpatterns_reverse/tests.py]: I think my previous comment got lost: Could you add "... concurrently but not more than once per thread ..." please.

**Pull request merged**